model/src/main/kotlin/Scope.kt

```diff
@@ -90,4 +91,42 @@ data class Scope(
 
         return getTreeDepthRec(dependencies)
     }
+
+    /**
+     * Return the shortest path to each dependency in this scope. The path to a dependency is defined by the nodes of
+     * the dependency tree that need to be passed to get to the dependency. For direct dependencies the shortest path is
+     * empty.
+     */
+    @JsonIgnore
+    fun getShortestPaths(): Map<Identifier, List<Identifier>> {
+        data class QueueItem(
+            val pkgRef: PackageReference,
+            val parents: List<Identifier>
+        )
+
+        val remainingIds = collectDependencies().toMutableSet()
+        val queue = LinkedList<QueueItem>()
+        val result = sortedMapOf<Identifier, List<Identifier>>()
+
+        dependencies.forEach { queue.offer(QueueItem(it, emptyList())) }
+
+        while (queue.isNotEmpty()) {
+            val item = queue.poll()
+            if (item.pkgRef.id in remainingIds) {
```

Comment:
If a queue item with an equal `id` has already been dequeue, nothing should be done except for `continue` to the next iteration, that is the execution of lines 120 - 123 is not needed and should not happen.

Reply:
That optimization would only be correct if the subtrees of a package were guaranteed to be identical, but they are not due to version conflict resolution. For example you could have a tree like this:
```
- A:1.0
  - B:1.0
    - C:1.0 (defined by version range that allows 1.x)
- D:1.0
  - B:1.0
    - C:1.1
  - E:1.0
    -F:1.0
      - C:1.1
```

In this case the shortest path `D:1.0->B:1.0` to `C:1.1` would not be found with that optimization, but instead `D:1.0->E:1.0->F:1.0`.

Reply:
Oh you're right. I missed that completely. Thanks for the explanations.